### ISSUE TITLE
Which options are "input options"?

### ISSUE THREAD

**Issue by cwebber** [w3c-dvcg/ld-signatures#13]
The [Create Verify Hash Algorithm](https://w3c-dvcg.github.io/ld-signatures/#create-verify-hash-algorithm) mentions "input options", distinct from signature options apparently, seemingly out of nowhere.  It turns out these options are a subset of the possible options available to the signature/verify algorithms, and on verify in fact appear to come from the key and signature of the object being verified mostly itself.  However nowhere else in the document is this mentioned, and which keys to extract, and from where, is not spelled out.

In practice from jsonld-signatures.js we can see that they are the following:

``` javascript
      {
        algorithm: algorithm,
        privateKeyPem: privateKeyPem,
        privateKeyWif: privateKeyWif,
        date: date,
        nonce: nonce,
        domain: domain
      }
```

We should specify how to extract these from what in the verify/signature sections when explaining that they must be passed into Create Verify Hash.  It may also be that some of these values may be different depending on the signature suite used, so we should specify some way that the suite can determine how/where to "extract" these values.

**Comment by msporny** [comment]
> mentions "input options", distinct from signature options apparently, seemingly out of nowhere.

Hmm, may need to rephrase that section. `signature options` is the concept... `input options` is the name of the variable used in the algorithm. They're the same thing, the latter is the concrete name for the former. Does that make sense? If so, we can figure out how to rephrase it so it's more clear.

> However nowhere else in the document is this mentioned, and which keys to extract, and from where, is not spelled out.

Yes it's not spelled out, mostly because we were trying to not be too prescriptive to implementers. I'm happy to add something more specific, as long as we're not telling implementers exactly how to pass parameters around and from where they should get those parameters. It's fine to give hints and notes, though. What text would you find helpful?

> We should specify how to extract these from what in the verify/signature sections when explaining that they must be passed into Create Verify Hash.

Got some proposed spec text?

>  It may also be that some of these values may be different depending on the signature suite used, so we should specify some way that the suite can determine how/where to "extract" these values.

Again, this is up to implementers to decide and we should be careful about overspecifying. In general, any signature suite specific stuff is passed in via the `options`. When the LDS spec was written, we didn't have the Koblitz curve signatures specified, and now that they are, perhaps we can look for generalizations there.

In a brief glance, I think we can perhaps rename `input options` to just `options`. That may remove the misinterpretation you had of the spec.

**Comment by cwebber** [comment]
> Hmm, may need to rephrase that section. signature options is the concept... input options is the name of the variable used in the algorithm. They're the same thing, the latter is the concrete name for the former. Does that make sense? If so, we can figure out how to rephrase it so it's more clear.
> In a brief glance, I think we can perhaps rename input options to just options. That may remove the misinterpretation you had of the spec.

That makes sense, and I think using the same term of `options` will reduce confusion about whether there are multiple things.

As background, one of the reasons for my confusion (on this and several of the other issues filed) was I was trying to figure out whether or not I could make the api more "pythonic" by using python-style keyword arguments instead of an `options` argument (which is also what's used in pyld).  I started down that path then turned around once I saw this line:

> _Creating a canonicalized options document by canonicalizing options according to the canonicalization algorithm (e.g. the GCA2015 [RDF-DATASET-NORMALIZATION] algorithm)._

At this point I figured, oh, I have to actually normalize this object, so I guess I should leave it as a hashmap.  Clearly that doesn't make sense now though, since some of the values would be procedures, and there's no way to normalize and attach that to the signature.  With all that in retrospect the source of my confusion (from comparing the spectext and reference code side by side) was between conflating `options` as an argument to a procedure (which the procedure is given a lot of leeway to) and `options` as in options that apply to the signature suite and which will be normalized and attached to the signature (which includes some standard ones like `nonce`).  That distinction seems a lot clearer now (assuming I did get it).

> Got some proposed spec text?

I am thinking about it and will come back to it.  I'm also seeing where some of my confusion came from was misreading commas as indicating separate arguments and conflating different meanings of `options` as discussed above.  Perhaps nobody else would be confused by this, but I'll take a shot at writing up some text I would find clearer and make a PR.

**Comment by msporny** [comment]
> Perhaps nobody else would be confused by this

No, I think other people are going to get tripped up there as well. We need to figure out a different way of marking and talking about the variables. Maybe using `<code>options</code>`... or something that italicizes and color codes. Problem there being people that print the spec out as black and white.
